tracing/tracing.sh

```diff
@@ -31,3 +31,145 @@ function sync_main_kw_file_with_tracing()
+      file_with_tracing+='{'$'\n'
+      file_with_tracing+="$(get_tracing_log_line 'entry' "$function_name")"$'\n'
+      read -r line
+    else
+      file_with_tracing+="$line"$'\n'
+    fi
+
+    while [[ -n "$inside_function" ]]; do
+      read -r line
+      # End of function definition
+      if [[ "$line" == '}' ]]; then
+        file_with_tracing+='local _return_val="$?"'$'\n'
+        file_with_tracing+="$(get_tracing_log_line 'return' "$function_name")"$'\n'
+        file_with_tracing+='return "$_return_val"'$'\n'
+        file_with_tracing+='}'$'\n'
+        inside_function=''
+      # Handle single-line return statement
+      elif [[ "$line" =~ ^[[:space:]]*return ]]; then
+        file_with_tracing+="$(get_tracing_log_line 'return' "$function_name")"$'\n'
+        file_with_tracing+="$line"$'\n'
+      # Handle statement with '&& return'
+      elif [[ "$line" =~ \&\&[[:space:]]return ]]; then
+        file_with_tracing+="${line%"$and_return_statement"*}&& "
+        file_with_tracing+="$(get_tracing_log_line 'return' "$function_name") "
+        file_with_tracing+="${and_return_statement}${line#*"$and_return_statement"}"$'\n'
+      # Handle single-line exit statement
+      elif [[ "$line" =~ ^[[:space:]]*exit ]]; then
+        file_with_tracing+="$(get_tracing_log_line 'exit' "$function_name")"$'\n'
+        file_with_tracing+="$line"$'\n'
+      # Handle statements with '&& exit'
+      elif [[ "$line" =~ \&\&[[:space:]]exit ]]; then
+        file_with_tracing+="${line%"$and_exit_statement"*}&& "
+        file_with_tracing+="$(get_tracing_log_line 'exit' "$function_name") "
+        file_with_tracing+="${and_exit_statement}${line#*"$and_exit_statement"}"$'\n'
+      # Handle background execution
+      elif [[ "$line" =~ [^\&\;]\&$ ]]; then
+        file_with_tracing+="THREAD_NUMBER=\$((++TOTAL_NUMBER_OF_THREADS))"$'\n'
+        file_with_tracing+="$line"$'\n'
+        file_with_tracing+="THREAD_NUMBER=0"$'\n'
+      else
+        file_with_tracing+="$line"$'\n'
+      fi
+    done
+  done < "$filepath"
+
+  printf '%s' "$file_with_tracing"
+}
+
+# This function outputs an expression that logs a tracing event when it gets
+# evaluated. A tracing event can be either entering (entry) or returning (return)
+# into/from a function, or exiting (exit) the execution. When evaluated, the
+# outputted expression adds a CSV line to the current thread tracing file with
+# the format:
+#
+#   <tracing_event>,<function_name>,<timestamp>
+#
+# @tracing_event: String that represent the entry/return of a function, or the
+#   exit of the execution
+# @function_name: Name of the function of tracing event
+#
+# Return:
+# Outputs an expression ready to be evaluated.
+function get_tracing_log_line()
+{
+  local tracing_event="$1"
+  local function_name="$2"
+  local expression
+
+  expression="printf '${tracing_event},${function_name},%s\n' "
+  expression+='"$(date '+%s%N')" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'
```

Comment:
Maybe we should consider adding a proper function for this date in the lib file?

Reply:
Hmm, I see the point, but I'm not sure that this would be a better approach.

As this string `"$(date '+%s%N')"` isn't being evaluated in this function, but rather will be evaluated in any other function defined in `src`, if we substitute this string for a call to a function, say, `get_timestamp_in_nano_secs` defined in `src/lib/kw_date_and_time.sh` we would have to assure that every lib file included `kw_date_and_time.sh`.

Does my point makes sense?